runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/JobGroupTest.java

```diff
@@ -1293,6 +1309,88 @@ public IStatus run(IProgressMonitor monitor) {
 		assertTrue("2.0", numShouldCancelCalled[0] < NUM_JOBS_LIMIT + 10);
 	}
 
+	/**
+	 * Tested scenario: - JobGroup.shouldCancel returns true after a single job is
+	 * canceled.
+	 *
+	 * Expected result: The remaining jobs are canceled in a reasonable time after
+	 * the shouldCancel method of the JobGroup returns true. Some jobs might even
+	 * have to be explicitly asked to cancel their execution via the method
+	 * "canceling"
+	 */
+	public void testShouldCancel_5() {
+		final int MAX_THREADS = 10;
+
+		// make sure some jobs haven't yet been queued by the time the cancelation takes
+		// place
+		final int NUM_JOBS_LIMIT = MAX_THREADS + 100;
+		final int NUM_JOBS = NUM_JOBS_LIMIT + 1000;
+
+		// counters
+		final int countJobsRan[] = { 0 };
+		final int countJobsCanceled[] = { 0 };
```

Comment:
I would prefer to use `AtomicInteger` or the like, as code becomes pretty hard to read when abusing an array with a single element named after the meaning of the first element rather than the meaning of the array.

Reply:
In addition, incrementing the first array element from different threads is a race condition.